File: ex343.java
class Solution {
    public int integerBreak(int n) {
        int[] dp = new int[n + 1];
        
        for (int i = 0; i <= n; i++) {
            for (int j = 0; j <= i; j++) {
                dp[i] = Math.max(dp[i], j * dp[i - j]);
                dp[i] = Math.max(dp[i], j * (i - j));
            }
        }   
        
        return dp[n];
    }
}

File: code/ex343.cpp
class Solution {
public:
    int integerBreak(int n) {
        vector<int> dp(59);
        dp[1] = 1;
        
        for (int i = 2; i <= n; i++) {
            for (int n1 = 1; n1 < i; n1++) {
                int n2 = i - n1;
                dp[i] = max(dp[i], dp[n1] * dp[n2]);
            }
            if (i == n)    return dp[i];
            dp[i] = max(dp[i], i);
        }
        
        return dp[n];
    }
};